feat: Implement header-based collection and geometry filtering

bountx opened the pull request:
**Related Issue(s):**

- #516

**Description:**
Adds support for filtering search results based on HTTP headers. This enables granular access control at the proxy layer while keeping sfeos authorization-agnostic. 

Headers:
`X-Filter-Collections`: Comma-separated collection IDs to restrict access
`X-Filter-Geometry`: GeoJSON geometry for spatial filtering (requires optional dependency `shapely`)

Affected endpoints:
| Endpoint                           | Collection Filter   | Geometry Filter     |
|------------------------------------|---------------------|---------------------|
| `GET /collections`                 | Yes                 | Doesn't apply       |
| `GET /collections/{id}`            | Yes (404 if denied) | Doesn't apply       |
| `GET /collections/{id}/items`      | Yes (404 if denied) | Yes (via search)    |
| `GET /collections/{id}/items/{id}` | Yes (404 if denied) | Yes (404 if denied) |
| `GET/POST /search`                 | Yes                 | Yes                 |

Important:
There are no overhead computations if headers are not present for backwards compatibility.

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] Changes are added to the changelog

Gomez324 committed:
CAT-1395 - CI/CD

Gomez324 committed:
test

Gomez324 committed:
fix

Gomez324 committed:
fix

Gomez324 committed:
test 2

Gomez324 committed:
final?

Gomez324 committed:
fix

Gomez324 committed:
fix

Gomez324 committed:
fix

Gomez324 committed:
ci/cd

Gomez324 committed:
update gitlab-ci.yml

Gomez324 committed:
fix

Gomez324 committed:
remove tests from the pipeline

Gomez324 committed:
remove tag from the pipeline

jonhealy1 commented on stac_fastapi/core/stac_fastapi/core/core.py:
```diff
@@ -844,6 +887,19 @@ async def post_search(
+            search = self.database.apply_intersects_filter(
+                search=search, intersects=geometry_obj
+            )
+
```
Let's try to remove some code from core.py ie. 
```
if header_geometry := parse_filter_geometry(request):
    geometry_obj = create_geometry_filter_object(header_geometry)
    search = self.database.apply_intersects_filter(
        search=search, intersects=geometry_obj
    )

jonhealy1 replied:
This too:
```
# Check if collection is allowed by header filter
header_collections = parse_filter_collections(request)
if header_collections is not None and collection_id not in header_collections:
    raise HTTPException(status_code=404, detail="Collection not found")
  ```
  
Moving these blocks cleans up core.py significantly by removing:

1. Local imports (from types import ..., from ...header_filters)
2. try/except or if/raise permission logic
3. Data structure instantiation (SimpleNamespace)

jonhealy1 reviewed (changes requested):
@bountx Looks really good. Let's just clean up core a little. Thank you

jonhealy1 reviewed (dismissed):
@bountx Thanks

jonhealy1 reviewed (changes requested):
This looks great, but there is a significant performance pitfall when mixing standard `datetime` parameters with CQL2 filters.

Currently, if a request has `datetime="..."` in the body but the CQL2 filter does not contain a date, `resolve_cql2_indexes` ignores the body datetime. This forces the system to scan **all** indices instead of just the relevant time partition.

**Suggested Fix:**
Please update `resolve_cql2_indexes` to accept `datetime_search` as a fallback argument.

In `stac_fastapi/sfeos_helpers/filter/cql2.py`:
```python
        if date_range:
            _, collection_datetime = apply_datetime_filter(
                search, format_datetime_range(date_str=date_range)
            )
        elif datetime_search:
            # Fallback to standard parameter
            collection_datetime = datetime_search

jonhealy1 reviewed (dismissed):
Thanks @bountx

z-mrozu committed:
[CAT-1504] waw3-2...-staging gitlab ci update

Gomez324 committed:
Fixed end_datetime selection when splitting indices in a datetime-based indexing strategy.

Gomez324 committed:
pre-commit

Gomez324 committed:
update docker

Gomez324 committed:
update docker

Gomez324 committed:
update docker

Gomez324 committed:
fix

Gomez324 committed:
ops without dockerfile from github

Gomez324 committed:
sentry

Commit:
Resolve CAT-1672 "Ops dep"

z-mrozu committed:
optional deploy-dev

Commit:
feat: implement AST structure for cql2-json queries index selection

- Added Abstract Syntax Tree (AST) representation for CQL2 queries replacing dictionary-based queries
- Implemented datetime extraction logic to identify datetime, start_datetime, and end_datetime
- Support for complex queries with logical operators, spatial operations, and property comparisons
- Improved query parsing and maintainability of CQL2-JSON filter requests

Commit:
feat: implement AST structure for cql2-json queries index selection

- Added Abstract Syntax Tree (AST) representation for CQL2 queries replacing dictionary-based queries
- Implemented datetime extraction logic to identify datetime, start_datetime, and end_datetime
- Support for complex queries with logical operators, spatial operations, and property comparisons
- Improved query parsing and maintainability of CQL2-JSON filter requests

Commit:
test: add datetime index filtering tests for CQL2 operators

Add test coverage for filtering with CQL2 operators, ensuring proper index selection and query results
for POST and GET requests.

- test_search_request_cql2_equal_operator: datetime equal filtering with index selection
- test_search_request_cql2_greater_than_operator: greater-than datetime filtering excludes non-matching indexes
- test_search_request_cql2_less_than_operator: less-than datetime filtering excludes non-matching indexes
- test_search_request_cql2_between_operator: datetime between filtering excludes
- test_search_request_cql2_in_operator: IN filtering with datetime index selection
- test_search_request_cql2_not_equal_operator: NOT filtering with proper index selection
- test_search_request_cql2_and_operator: AND operator with multiple conditions and index selection
- test_search_request_cql2_or_operator: OR operator returns multiple items while selecting index
- test_search_request_cql2_not_operator: NOT operator for complex query excludes specific items
- test_search_request_cql2_like_operator: Test pattern matching with LIKE operator returns matching items
- test_search_request_cql2_is_null: IS NULL returns items missing the specified field
- test_search_request_cql2_is_not_null: IS NOT NULL returns items with the field present

Commit:
test: add datetime index filtering tests for CQL2 operators

Add test coverage for filtering with CQL2 operators, ensuring proper index selection and query results
for POST and GET requests.

- test_search_request_cql2_equal_operator: datetime equal filtering with index selection
- test_search_request_cql2_greater_than_operator: greater-than datetime filtering excludes non-matching indexes
- test_search_request_cql2_less_than_operator: less-than datetime filtering excludes non-matching indexes
- test_search_request_cql2_between_operator: datetime between filtering excludes
- test_search_request_cql2_in_operator: IN filtering with datetime index selection
- test_search_request_cql2_not_equal_operator: NOT filtering with proper index selection
- test_search_request_cql2_and_operator: AND operator with multiple conditions and index selection
- test_search_request_cql2_or_operator: OR operator returns multiple items while selecting index
- test_search_request_cql2_not_operator: NOT operator for complex query excludes specific items
- test_search_request_cql2_like_operator: Test pattern matching with LIKE operator returns matching items
- test_search_request_cql2_is_null: IS NULL returns items missing the specified field
- test_search_request_cql2_is_not_null: IS NOT NULL returns items with the field present

Commit:
docs: update readme & changelog

Commit:
docs: update readme & changelog

Commit:
feat: implement header-based filtering for collections and geometry

Commit:
refactor: simplify geometry intersection logic and improve error handling

Commit:
feat: add shapely as a dependency for Elasticsearch and Opensearch

Commit:
fix: add catalogue install to cicd

Commit:
tests: CQL2 datetime and collection index resolver

YuriZmytrakov committed:
Merge branch 'main' into CAT-1477

bountx committed:
Merge branch 'main' into CAT-1624

Commit:
Merge branch 'CAT-1477' into CAT-1624

Commit:
fix cql2_metadata wrong init

Commit:
Merge branch 'ops' into CAT-1624

Gomez324 committed:
Ops

Gomez324 committed:
Merge branch 'ops' into 'main'

Ops

Closes CAT-1672

See merge request stac/sfeos!3

Gomez324 committed:
Merge branch 'CAT-1868' into 'master'

CAT-1868 - The datetime indexing method has been removed

Closes CAT-1868

See merge request stac/sfeos!28

Gomez324 committed:
CAT-1868 - The datetime indexing method has been removed

Gomez324 committed:
Resolve CAT-1924, CAT-1934

Gomez324 committed:
Merge branch 'CAT-1924' into 'master'

Resolve CAT-1924, CAT-1934

Closes CAT-1924

See merge request stac/sfeos!30

Gomez324 committed:
CAT-1937 - Added missing tag_check job to the pipeline.

Gomez324 committed:
Merge branch 'CAT-1937' into 'master'

CAT-1937 - Added missing tag_check job to the pipeline.

Closes CAT-1937

See merge request stac/sfeos!31

Commit:
Merge branch 'CAT-1728' into 'master'

[CAT-1728] Add type validation for fields in mapping

See merge request stac/sfeos!38

Commit:
[CAT-1728] Add type validation for fields in mapping

Gomez324 committed:
CAT-1909, CI/CD SFEOS

Gomez324 committed:
Merge branch 'CAT-1909' into 'master'

CAT-1909, CI/CD SFEOS

Closes CAT-1909

See merge request stac/sfeos!26

Gomez324 committed:
Merge branch 'CAT-1836' into 'master'

Resolve CAT-1836

Closes CAT-1836

See merge request stac/sfeos!29

Gomez324 committed:
Resolve CAT-1836

Commit:
Add support for collection blacklisting in header filters

Commit:
pre-commit

Commit:
Add collection exclusion filter to database logic

Commit:
Merge branch 'master' into CAT-1624

Commit:
Refactor datetime string formatting in extract_collection_datetime function

Commit:
Fix environment name options in deploy script

Commit:
Add collection filtering support in database logic for Elasticsearch and OpenSearch

Commit:
[CAT-1926] Add De Morgan's law tests and closed index aliases for CQL2

Commit:
Merge branch 'CAT-1926' into 'master'

[CAT-1926] Add De Morgan's law tests and closed index aliases for CQL2

See merge request stac/sfeos!36

Commit:
Merge branch 'master' into CAT-1624
